sip/sip-009-analysis-cost-assessment.md

```diff
@@ -0,0 +1,333 @@
+from the block.
+
+Costs for static analysis are assessed during the _type check_ pass.
+The read-only and trait-checking passes perform work which is strictly
+less than the work performed during type checking, and therefore, the
+cost assessment can safely fold any costs that would be incurred during
+those passes into the type checking pass.
+
+# Common Analysis Metrics and Costs
+
+## AST Parsing
+
+The Clarity parser has a runtime that is linear with respect to the Clarity
+program length.
+
+```
+a*X+b
+```
+
+where a and b are constants, and
+
+X := the program length in bytes
+
+## Dependency cycle detection
+
+Clarity performs cycle detection for intra-contract dependencies (e.g.,
+functions that depend on one another). This detection is linear in the
+number of dependency edges in the smart contract:
+
+```
+a*X+b
+```
+
+where a and b are constants, and
+X := the total number of dependency edges in the smart contract
+
+Dependency edges are created anytime a top-level definition refers 
+to another top-level definition.
+
+## Type signature size
+
+Types in Clarity may described using type signatures. For example,
+`(tuple (a int) (b int))` describes a tuple with two keys `a` and `b`
+of type `int`. These type descriptions are used by the Clarity analysis
+passes to check the type correctness of Clarity code. Clarity type signatures
+have varying size, e.g., the signature `int` is smaller than the signature for a
+list of integers.
+
+The signature size of a Clarity type is defined as follows:
+
+```
+type_signature_size(x) :=
+  if x = 
+     int      => 1
+    uint      => 1
+    bool      => 1
+    principal => 1
+    buffer    => 2
+    optional  => 1 + type_signature_size(entry_type)
+    response  => 1 + type_signature_size(ok_type) + type_signature_size(err_type)
+    list      => 2 + type_signature_size(entry_type)
+    tuple     => 1 + 2*(count(entries)) 
+                   + sum(type_signature_size for each entry)
+                   + sum(len(key_name) for each entry)
+```
+
+## Type annotation
+
+Each node in a Clarity contract's AST is annotated with the type value
+for that node during the type checking analysis pass.
+
+The runtime cost of type annotation is:
+
+```
+a + b*X
+```
+
+where a and b are constants, and X is the type signature size of the
+type being annotated.
+
+## Variable lookup
+
+Looking up variables during static analysis incurs a non-constant cost -- the stack
+depth _and_ the length of the variable name affect this cost. However,
+variable names in Clarity have bounded length -- 128 characters. Therefore,
+the cost assessed for variable lookups may safely be constant with respect
+to name length.
+
+The stack depth affects the lookup cost because the variable must be
+checked for in each context on the stack.
+
+Cost Function:
+
+```
+a*X+b*Y+c
+```
+
+where a, b, and c are constants,
+X := stack depth
+Y := the type size of the looked up variable
+
```

Comment:
As symbols cannot be shadowed, it shouldn't strictly be necessary to search the stack for variables, by instead maintaining them in a global map for constant time lookups.

Reply:
Yep, that's true. For now, both the runtime and the static analysis use context nesting for variable lookups, but this should absolutely be revisited, if only because it will simplify the operation costs.

Reply:
I started an issue to track this: https://github.com/blockstack/blockstack-core/issues/1325